# Import libraries

In [134]:
# import Libraries
import pandas as pd
import numpy as np
import os

# import dataframes

In [135]:
#Identify path 
path= '../02 Data/'

In [136]:
# import df_ords_prods_merge pickle
ords_prods_merge=pd.read_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge.pkl'))

# Task Procedures

## aggreagte mean

In [137]:
# getting mean fr entire dataframe; by order number by department id
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## answer:
the results of the mean for the whole dataset varied a bit. not too much, but i did notice some rows had a slightly high or lower mean. this makes senes that hte intrpduction of the rest of rows will alter the means. for exmpletdept_id 21 in the subset had a mean of 25.53 but in the full analysis has a mean o 22.90

In [138]:
# loyalty flag for customers with transform () and loc()
# create a loyalty flag using loc()- create ' max order'
ords_prods_merge['max_order']= ords_prods_merge.groupby (['user_id']) ['order_number'].transform ('max')

In [139]:
ords_prods_merge.head(5)

product_id                             product_name  aisle_id  \
92368            196                                     Soda        77   
7157228        12427                      Original Beef Jerky        23   
8435031        14084  Organic Unsweetened Vanilla Almond Milk        91   
16642084       26088               Aged White Cheddar Popcorn        23   
17009387       26405         XL Pick-A-Size Paper Towel Rolls        54   

          department_id  prices  order_id  user_id  order_number  \
92368                 7     9.0   2539329        1             1   
7157228              19     4.4   2539329        1             1   
8435031              16    12.5   2539329        1             1   
16642084             19     4.7   2539329        1             1   
17009387             17     1.0   2539329        1             1   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
92368                      2                  8                     NaN   
7157228                    2                  8                     NaN   
8435031                    2                  8                     NaN   
16642084                   2                  8                     NaN   
17009387                   2                  8                     NaN   

          First_order  add_to_cart_order  reordered _merge  \
92368            True                  1          0   both   
7157228          True                  3          0   both   
8435031          True                  2          0   both   
16642084         True                  4          0   both   
17009387         True                  5          0   both   

             price_range_loc    busiest_days busiest_period_of_day  max_order  \
92368     Mid range products  Regularly busy        Average orders         10   
7157228   Low range products  Regularly busy        Average orders         10   
8435031                  NaN  Regularly busy        Average orders         10   
16642084  Low range products  Regularly busy        Average orders         10   
17009387  Low range products  Regularly busy        Average orders         10   

          loyalty_flag  
92368     New customer  
7157228   New customer  
8435031   New customer  
16642084  New customer  
17009387  New customer

## loyalty flag using loc() function


In [140]:
#If the max orders the user has made is over 40, then the customer will be labeled a ‘loyal customer’
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag']= 'Loyal customer'

In [141]:
#In the max orders the user has made is over but less than or equal to 40, then the customer will be labeled a ‘ regular customer’ 
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag']= 'Regular customer'


In [142]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag']= 'New customer'

In [143]:
#checking frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [152]:
# Replace prices > 100 with NaN // getting rid of outliers
ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'] = np.nan

In [153]:
# compare the prices for loyalty column
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max', 'std']})

prices                     
                      mean  min   max       std
loyalty_flag                                   
Loyal customer    7.773575  1.0  25.0  4.206151
New customer      7.801206  1.0  25.0  4.266834
Regular customer  7.798262  1.0  25.0  4.254886

# answer:
we are able to see that Loyal customers average spending to be $7.78 compared to New customers at $7.80 and Regualr customers at $7.80. new customers tend to spend slighting more than loyal customers 

# create a spending flag

In [154]:
#aggregating data with transformed ()
ords_prods_merge['avg_product_price']= ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [197]:
ords_prods_merge[['user_id','avg_product_price', 'spending_flag', 'Median_days_between']].head(15)

user_id  avg_product_price spending_flag  Median_days_between
92376           1           6.367797   Low spender                 20.5
5971722         1           6.367797   Low spender                 20.5
7157236         1           6.367797   Low spender                 20.5
16105131        1           6.367797   Low spender                 20.5
29712648        1           6.367797   Low spender                 20.5
32034007        1           6.367797   Low spender                 20.5
92375           1           6.367797   Low spender                 20.5
5971721         1           6.367797   Low spender                 20.5
7157235         1           6.367797   Low spender                 20.5
16105130        1           6.367797   Low spender                 20.5
29712647        1           6.367797   Low spender                 20.5
32034006        1           6.367797   Low spender                 20.5
92369           1           6.367797   Low spender                 20.5
5971715         1           6.367797   Low spender                 20.5
7157229         1           6.367797   Low spender                 20.5

In [156]:
# Sort by user_id to see the grouping, and order_number to see the sequence
ords_prods_merge = ords_prods_merge.sort_values(by=['user_id', 'avg_product_price'])

In [168]:
#if the mean of the price of prodcuts purchased by a user is lower than 10, then flag as a 'low spender'
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag']= 'Low spender'

In [169]:
#if the mean of the prices of products purchased by a user is higher than or equal to 10, then flag as 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag']= 'High spender'

##  Frequent customers column

In [190]:
#aggregating data with transformed ()
ords_prods_merge['Median_days_between']= ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [191]:
# Sort by user_id to see the grouping, and order_number to see the sequence
ords_prods_merge = ords_prods_merge.sort_values(by=['user_id', 'days_since_prior_order'])

In [192]:
# if the median of "days since prior order' is higher than 20, then the customer is 'non-frequent customer'
ords_prods_merge.loc[ords_prods_merge['Median_days_between'] >20, 'frequent_flag']= 'non-frequent customer'

In [193]:
#If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
ords_prods_merge.loc[(ords_prods_merge['Median_days_between'] >10) & (ords_prods_merge['frequent_cust'] <=20), 'frequent_flag']= 'Regular customer'

In [194]:
#If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”
ords_prods_merge.loc[ords_prods_merge['Median_days_between'] <=10, 'frequent_flag']= 'frequent customer'

In [195]:
#checking frequency
ords_prods_merge['frequent_flag'].value_counts(dropna=False)

frequent_flag
frequent customer        21559853
Regular customer          7208564
non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [196]:
ords_prods_merge.head()

product_id           product_name  aisle_id  department_id  prices  \
92376            196                   Soda        77              7     9.0   
5971722        10258             Pistachios       117             19     3.0   
7157236        12427    Original Beef Jerky        23             19     4.4   
16105131       25133  Organic String Cheese        21             16     8.6   
29712648       46149      Zero Calorie Cola        77              7    13.4   

          order_id  user_id  order_number  orders_day_of_week  \
92376      2295261        1             9                   1   
5971722    2295261        1             9                   1   
7157236    2295261        1             9                   1   
16105131   2295261        1             9                   1   
29712648   2295261        1             9                   1   

          order_hour_of_day  ...     price_range_loc  busiest_days  \
92376                    16  ...  Mid range products  Busiest days   
5971722                  16  ...  Low range products  Busiest days   
7157236                  16  ...  Low range products  Busiest days   
16105131                 16  ...  Mid range products  Busiest days   
29712648                 16  ...  Mid range products  Busiest days   

          busiest_period_of_day  max_order  loyalty_flag avg_product_price  \
92376            Average orders         10  New customer          6.367797   
5971722          Average orders         10  New customer          6.367797   
7157236          Average orders         10  New customer          6.367797   
16105131         Average orders         10  New customer          6.367797   
29712648         Average orders         10  New customer          6.367797   

         spending_flag frequent_cust          frequent_flag  \
92376      Low spender          20.5  non-frequent customer   
5971722    Low spender          20.5  non-frequent customer   
7157236    Low spender          20.5  non-frequent customer   
16105131   Low spender          20.5  non-frequent customer   
29712648   Low spender          20.5  non-frequent customer   

         Median_days_between  
92376                   20.5  
5971722                 20.5  
7157236                 20.5  
16105131                20.5  
29712648                20.5  

[5 rows x 25 columns]

In [ ]:
# getting rid of outliers 
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

#### fixing Nan issue for price_range_loc column

In [148]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc']= 'High range product'

In [149]:
ords_prods_merge.loc[(ords_prods_merge['prices']<=15) & (ords_prods_merge['prices']>5), 'price_range_loc']= 'Mid range products'

In [150]:
ords_prods_merge.loc[ords_prods_merge['prices'] <=5, 'price_range_loc']='Low range products'

In [151]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid range products    21860860
Low range products    10126321
High range product      417678
Name: count, dtype: int64

# Export

In [200]:
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge.pkl'))